### PR TITLE
fix: log events missing on boot time

### PR DESCRIPTION
**Pull request by ilasw**
## Changes
Main change are: 
- `isInitialFetch` in BootProvider is now a computed boolean from query, preventing delays and possible bugs;
- moved some `useRef` to simpler implementations, using useCallback to reducing updates and make dependency arrays more stable;
- moving log functions to indipendent `useCallback` reducing the amount of updates fired from them;

## Events

Did you introduce any new tracking events?

## Experiment

Did you introduce any new experiments?

## Manual Testing

> [!CAUTION]
> Please make sure existing components are not breaking/affected by this PR

### Preview domain
https://fix-analytics-boot.preview.app.daily.dev

### CODE REVIEW

File: packages/shared/src/contexts/BootProvider.tsx [hunk 1]
```diff
@@ -84,12 +84,8 @@ const updateLocalBootData = (
   return result;
 };
 
-const getCachedOrNull = () => {
-  try {
-    return JSON.parse(storage.getItem(BOOT_LOCAL_KEY));
-  } catch (err) {
-    return null;
-  }
+const getCachedBootOrNull = () => {
```
Comment (idoshamun) [inline packages/shared/src/contexts/BootProvider.tsx:87]: Why did you change this implementation? I think we have the try/catch as in some cases `storage.getItem` may fail

File: packages/shared/src/hooks/log/useLogContextData.ts [hunk 2]
```diff
@@ -51,33 +51,42 @@ export default function useLogContextData(
...
-        const event = durationEventsQueue.current.get(id);
-        if (event) {
-          durationEventsQueue.current.delete(id);
-          event.event_duration =
-            now.getTime() - event.event_timestamp.getTime();
-          if (window.scrollY > 0 && event.event_name !== 'page inactive') {
-            event.page_state = 'active';
-          }
-          pushToQueue([event]);
+  const logEvent = useCallback(
```
Comment (idoshamun) [inline packages/shared/src/hooks/log/useLogContextData.ts:54]: What's the actual difference between `useCallback` for each function vs `useMemo` for all?

File: packages/shared/src/hooks/log/useLogPageView.ts [hunk 3]
```diff
@@ -1,32 +1,29 @@
-import { MutableRefObject, useContext, useEffect, useRef } from 'react';
+import { useCallback, useContext, useEffect } from 'react';
 import { useRouter } from 'next/router';
 import LogContext from '../../contexts/LogContext';
 
-export default function useLogPageView(): MutableRefObject<() => void> {
+export default function useLogPageView(): () => void {
```
Comment (idoshamun) [inline packages/shared/src/hooks/log/useLogPageView.ts:5]: We need to make sure that moving from ref to callback doesn't cause duplicate events to be triggered

File: packages/shared/src/hooks/log/useLogContextData.ts (hunk 2 above)
Comment (ilasw) [reply packages/shared/src/hooks/log/useLogContextData.ts:54]: With a unique useMemo we force to compute again for each of dependency changes, with independent useCallback we only update the function when the method-specific dependency changes. 

Additionally useCallback is better in this case because it saves specifically the function and not the whole object as result of the hook!

File: packages/shared/src/contexts/BootProvider.tsx (hunk 1 above)
Comment (ilasw) [reply packages/shared/src/contexts/BootProvider.tsx:87]: This is not mandatory, so I can revert!
I've changed this because maybe we don't need it. According to documentation getItem can return `null` if the key didn't exist. This is making JSON.stringify to throw errors because accepts only `string` values.

File: packages/shared/src/hooks/log/useLogContextData.ts (hunk 2 above)
Comment (idoshamun) [reply packages/shared/src/hooks/log/useLogContextData.ts:54]: Thanks for the explanation!

File: packages/shared/src/contexts/BootProvider.tsx (hunk 1 above)
Comment (idoshamun) [reply packages/shared/src/contexts/BootProvider.tsx:87]: I think in incognitio it will fail, you can check. if it works no issues

Comment (ilasw) [reply packages/shared/src/contexts/BootProvider.tsx:87]: It works! 👍

Comment (capJavert) [reply packages/shared/src/contexts/BootProvider.tsx:87]: I think its better that it throws and we catch whatever, localStorage can sometimes even not be available, let's minimize changes to actual fix of the issue, its easier to review.

**Review by capJavert — COMMENTED** [review]
Having trouble reviewing this, could you point to some exact spots in code that had issue with data logging? I appreciate optimizations as well, just hard to parse the changes from reviewer side.

**Comment by ilasw** [comment]
@capJavert a combined change to `isInitialFetch` and the `useRef`s to `useCallback` was the key changes for this fix.
They where not linked in other functions dependencies and this may lead to not run useEffects in the right moment!

File: packages/shared/src/contexts/BootProvider.tsx [hunk 4]
```diff
@@ -186,7 +184,7 @@ export const BootDataProvider = ({
   const updatedAtActive = user ? dataUpdatedAt : null;
   const updateQueryCache = useCallback(
     (updatedBootData: Partial<BootCacheData>, update = true) => {
-      const cachedData = getCachedOrNull() || {};
+      const cachedData = getCachedBootOrNull() ?? {};
```
Comment (sshanzel) [inline packages/shared/src/contexts/BootProvider.tsx:187]: This might result to error. It will not fallback if the value is `null` compared to using `||`.

File: packages/shared/src/contexts/BootProvider.tsx [hunk 5]
```diff
@@ -274,7 +274,7 @@ export const BootDataProvider = ({
         isFetched={isBootReady}
         isLegacyLogout={bootData?.isLegacyLogout}
         accessToken={bootData?.accessToken}
-        isPastRegistration={isInitialFetch.current}
+        isPastRegistration={isInitialFetch}
```
Comment (sshanzel) [inline packages/shared/src/contexts/BootProvider.tsx:277]: Can you check auth flows if they would still work as intended.
- Register using email
- Register using email but refresh the page on account details form.
- Register using provider
- Register using provider but refresh the page on account details form.

**Comment by sshanzel** [comment]
> a combined change to isInitialFetch

I'm not sure this would affect the events since this is only used to check whether the user should be redirected after not completing the registration.

**Review by sshanzel — COMMENTED** [review]
If there is a race condition where the events are triggered before boot is ready, a fix I can think of is to have queue for the events called while we are still missing boot data, then fill the events once available before sending.

Although, Ido mentioned that the events affected were events with `target_type` of `post` that is not an `impression. Which means, the events for interacting with the post page. With that, the fixes on use refs/callbacks might actually work since dependencies are now applied correctly.

File: packages/shared/src/contexts/BootProvider.tsx (hunk 1 above)
Comment (sshanzel) [reply packages/shared/src/contexts/BootProvider.tsx:88]: Yeah, during SSR I think there can be an issue.

File: packages/shared/src/contexts/BootProvider.tsx [hunk 6]
```diff
@@ -129,8 +130,7 @@ export const BootDataProvider = ({
     );
   };
 
-  const [initialLoad, setInitialLoad] = useState<boolean>(null);
-  const [cachedBootData, setCachedBootData] = useState<Partial<Boot>>(() => {
+  const initialData = useMemo(() => {
```
Comment (sshanzel) [inline packages/shared/src/contexts/BootProvider.tsx:133]: I think we don't have to make this a useMemo since we only need it at the beginning to hold some data for initial render.

**Review by sshanzel — COMMENTED** [review]
Just one thing from my side but don't have anything blocking. With that, we should rigorously test it. Once we get the approvals, feel free to send another message to the same thread so we can also test it on our own.

**Comment by ilasw** [comment]
According to slack thread I'm merging and test on prod as soon as deployed
